content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,151 @@
+---
+id: ternary-search
+title: Ternary Search
+author: Chongtian Ma
+prerequisites:
+  - /silver/binary-search
+redirects:
+  - /bronze/ternary-search
+  - /gold/ternary-search
+  - /plat/ternary-search
+description: Searching for a peak in a unimodal function.
+frequency: 1
+---
+
+<Resources>
+	<Resource
+		source="CP-Algorithms"
+		title="Ternary Search"
+		url="https://cp-algorithms.com/num_methods/ternary_search.html"
+		starred
+	/>
+</Resources>
+
+Ternary search is the equivalent of binary searching for the zero on the
+derivative function. The most basic function can be represented by the standard
+form of a parabola: $ax^2+bx+c$. Depending on if $a$ is positive or negative,
+you are searching for the local minimum or maximum of the function,
+respectively. Note that the function must be unimodal (i.e. change directions in exactly one place)! If not, you probably need to use [Convex Hull Trick](https://usaco.guide/plat/convex-hull-trick). 
+
+Unlike binary search where a checking function $f(x)$ only needs to return true
+or false, we need $f(x)$ to output an integer. Between the bounds of $[l, r]$,
+we need to compare the integer output of $f(m_1)$ with $f(m_2)$ where
+$l < m_1 < m_2 < r$ to determine how $l$ and $r$ should be moved.
+
+For example, if the function has a local maximum between $[l, r]$ and
+$f(m_1) < f(m_2)$, then we know the local maximum cannot be between $[l, m_1]$,
+so we continue searching in $[m_1, r]$. For a mini-proof, consider the following
+cases:
+
+- Let's denote the local maximum as $g$. If $m_1$ and $m_2$ are less than $g$,
```

Comment:
i thought we didn't know what g was

Reply:
asssume bro

Reply:
that's not what i meant
how can we know any of these conditions if g is unknown in the first place

Reply:
well we know g is an integer, so it has to fall within these conditions..

Reply:
this explanation is fine to me :/

Reply:
well which condition is it? we only know which one it is if we know g, right?
am i high

Reply:
The explanation is confusing. all we need to show is that if $f(m_1) < f(m_2)$ then the case $g\in [l, m_1)$ is impossible. Why are we considering other cases?